### PR TITLE
Tr/cache git provider

### PR DESCRIPTION
**Pull request by mrT23**
### **PR Type**
enhancement, bug fix

___

### **Description**
- Added `get_git_provider_with_context` function to retrieve Git provider instance with context awareness.
- Refactored multiple modules to use `get_git_provider_with_context` for consistent context handling.
- Extended `GitProvider` class with new abstract methods `get_files` and `get_incremental_commits`.
- Improved error handling and context initialization in various modules.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Add context-aware Git provider retrieval function.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/git_providers/__init__.py

<li>Added <code>get_git_provider_with_context</code> function to retrieve Git provider <br>instance with context awareness.<br> <li> Modified <code>get_git_provider</code> to include context handling.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-91af7b74c12181fd8aac8cd7ec8af9a01aed1adfb67eff4748e8f6a6ca3cc660">+32/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>git_provider.py</strong><dd><code>Extend GitProvider with new abstract methods.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/git_providers/git_provider.py

<li>Added <code>get_files</code> abstract method to <code>GitProvider</code> class.<br> <li> Added <code>get_incremental_commits</code> method to <code>GitProvider</code> class.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-52d45f12b836f77ed1aef86e972e65404634ea4e2a6083fb71a9b0f9bb9e062f">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>github_provider.py</strong><dd><code>Refactor GithubProvider for context-aware retrieval.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/git_providers/github_provider.py

<li>Refactored <code>GithubProvider</code> to use <code>get_git_provider_with_context</code>.<br> <li> Added <code>_get_incremental_commits</code> method.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-28979ec713529d6f619bcba0b9e6a3002d1164d47c70ccfea7f15306618a1a11">+11/-7</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Use context-aware Git provider in utils.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/git_providers/utils.py

<li>Modified <code>apply_repo_settings</code> to use <code>get_git_provider_with_context</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-314199df3610cf35a472e952294faf0f5fbb516a2bd217e891a510fd1a698bca">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>github_app.py</strong><dd><code>Refactor GitHub app server for context-aware Git provider.</code></dd></summary>
<hr>
      
pr_agent/servers/github_app.py

<li>Refactored various functions to use <code>get_git_provider_with_context</code>.<br> <li> Added context initialization for <code>git_provider</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-3fb3f174152732d1b69953c8bd81b8a0a30f0e42100dc626d932da8371851608">+20/-17</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions.py</strong><dd><code>Use context-aware Git provider in PR code suggestions.</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/tools/pr_code_suggestions.py

- Modified to use `get_git_provider_with_context`.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-b57ba775e741d6f80bc4f8154b71330c011dae0ac43f3d0197e785b3e6b7117b">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_description.py</strong><dd><code>Use context-aware Git provider in PR description.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/tools/pr_description.py

- Modified to use `get_git_provider_with_context`.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_reviewer.py</strong><dd><code>Use context-aware Git provider in PR reviewer.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/tools/pr_reviewer.py

<li>Modified to use <code>get_git_provider_with_context</code>.<br> <li> Added handling for incremental commits.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/984/files#diff-8e265068e189a06852605cc694b03e92b523b4f8162077a2f3455ced4cdad8dc">+5/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by okotek — APPROVED** [review]

File: pr_agent/git_providers/__init__.py [hunk 1]
```diff
@@ -28,3 +30,33 @@ def get_git_provider():
...
+        try:
+            provider_id = get_settings().config.git_provider
+            if provider_id not in _GIT_PROVIDERS:
+                raise ValueError(f"Unknown git provider: {provider_id}")
+            git_provider = _GIT_PROVIDERS[provider_id](pr_url)
+            if is_context_env:
+                context["git_provider"] = {pr_url: git_provider}
+            return git_provider
+        except Exception as e:
+            raise ValueError(f"Failed to get git provider for {pr_url}") from e
```
Comment (coditamar) [inline pr_agent/git_providers/__init__.py:62]: /ask explain this snippet

**Comment by mrT23** [comment]
Preparing review...

**Comment by mrT23** [comment]
**[Persistent review](https://github.com/Codium-ai/pr-agent/pull/984#issuecomment-2177869170)** updated to latest commit https://github.com/Codium-ai/pr-agent/commit/0c1331f77e8a8b10cd6e90c2d16074326d9ed67f

**Comment by mrT23** [comment]
Preparing review...

**Comment by mrT23** [comment]
**[Persistent review](https://github.com/Codium-ai/pr-agent/pull/984#issuecomment-2177869170)** updated to latest commit https://github.com/Codium-ai/pr-agent/commit/0c1331f77e8a8b10cd6e90c2d16074326d9ed67f

**Comment by mrT23** [comment]
Preparing review...

**Comment by mrT23** [comment]
**[Persistent review](https://github.com/Codium-ai/pr-agent/pull/984#issuecomment-2177869170)** updated to latest commit https://github.com/Codium-ai/pr-agent/commit/0c1331f77e8a8b10cd6e90c2d16074326d9ed67f

**Comment by mrT23** [comment]
/improve
--pr_code_suggestions.num_code_suggestions_per_chunk="6"
--pr_code_suggestions.focus_only_on_problems=false

**Pull request merged**